### ISSUE TITLE
role  spec

### ISSUE THREAD

**Issue by bcoca** [ansible/proposals#39]
# Proposal: role  specification

*Author*: Brian Coca <@bcoca>
*Date*: 2016/11/32

- Status: New
- Proposal type: core design
- Targeted Release: future release
- Estimated time to implement: 10 days

## Motivation

Better document required vars  and configuration for roles and add runtime verification

### Problems

Currently it is unclear what is required for a role as input and there is no validation unless role author liters it with 'asserts/ X is defined' or rescue blocks, a simpler way to define required vars and an 'upfront' error should avoid the assert/block 'verbosity' and/or failing a role mid task list leaving things in an uncertain state.

## Solution proposal

- Have a method of defining required vars (for documentation purposes defaults/main.yml should also be added as those are 'optional' vars.

- Validate config on load
- Validate vars on execution (meta: validate_role_vars)? 

Both to error out early and avoid bad states

- have ansible-doc 'know' about these vars and default vars, extending it's use to show role 'documentation' (might need it's own proposal).

meta/spec.yml
----
```
main:
  vars:
    myvar1:
         desc: this is a test var required for this role to test var validation
         prompt:  optional? for if we want to tie into a prompt facility
         type: string #(default?), allows us to validate not only var existence but correct typing
         choices: ['yes', 'no', 'maybe', 'after a few beers'] # optional enum for var values
         default: yes # this value overrides defaults/ from the role
    myvar2:
         desc: only required field for vars like this one, still a test
         required: True
  config:
    hash_behaviour: merge
other.yml: 
 # for alternate entry points 
 .... 
```
- a section per entry point, 'main' is probably only one present in most cases
- defaults/required: can be defined in spec, spec overrides vars/defaults.yml when duplicate

optional/future
============
- prompt/question field: in case we want to prompt or feed external tool to know what 'question' to present the user as prompt label.

**Comment by geerlingguy** [comment]
+1 to this—one typo fix though—I think you mean `myvar2:` and not `myvar:2`.

It would make it easier for roles where a default can't be universally set, but the role also wants to ensure the playbook has previously set the variable. Some roles currently bail with an error, others do an assert test... in any case, it's an inconsistent user experience.

**Comment by drybjed** [comment]
The idea is interesting but I see a few problems with this.

Are the variables in `vars/required.yml` "live", or used by the role during execution? If so, and if the variables are used in the role itself, it might lead to conflicts. What if I want to use `{{ myvar1.type }}` in a template somewhere, will that be specially handled by Ansible? Alternatively, how variables other than dictionaries would be documented?

If the variables are not "live", and instead are defined for example in `meta/vars.yml` file which is read by Ansible only for internal parsing, this quickly leads to data duplication. I suppose that if the users only listed required variables there, that could be somewhat consistent and easy to maintain. The bulk of default variables usually provide some default value, in my experience variables that are required to be sed by the user specifically are very rare.

In short, I would see this implemented not on the normal Ansible level where variables could be misinterpreted, but at a meta level, similar to `meta/main.yml` file.

**Comment by bcoca** [comment]
@drybjed the variables in required are not instantiated in the playbook, requried is only for doc/validation. That said, you would expect the variable to already exist in the play before it is used.

both examples are 'non dictionary' variables, not sure what you mean, a dict variable would look like:
```
myvar3: 
       desc: a test dictionary var.
       type: dict
```

As for location, meta/required_vars.yml  might be better, I'm not sure i want to stuff this in main.

**Comment by mscherer** [comment]
" Date: 2016/11/32" ?

And yes, I think keeping thing outside of vars/ is better. 

Since one goal is to be used for documentation, translation come to mind, but I wonder if that's not too much and out of scope for now.

**Comment by bcoca** [comment]
@mscherer you owe me a new keyboard ... spit my coffee on it when you made me think of role documentation i18n, i10n  and translations (gettext?) ....

**Comment by srvg** [comment]
+1 on putting required.yml in the meta folder, not the vars folder
I also like how a variable get's defied, templated. 
How about letting a variable also define if it should get overrruled, or hash merged? Perhaps also allowing a new var merge option: "listify" (as in a role that defines an application and that gets applied several times on one host: each app needs to be added up into a list of apps.)

**Comment by haad** [comment]
what would be the process to start working on this ? Is there a consensus that after implementing it could be included in a main repo ?

**Comment by bcoca** [comment]
@haad someone just needs to spend time on it, i think that we can get it approved quite quickly at an IRC meeting, I just don't have time to spare to implement it myself.

I would help out/guide/review if anyone else were to pick it up.

**Comment by privateip** [comment]
Cross posting this back from 134...

One thing to consider is that a given role could have more than one `survey.yml` file. With the introduction of `tasks_from` in `include_role` and `import_role`, it cannot be assumed that there is a 1:1 mapping between role and use case. It is absolutely conceivable that a single role could have more than one entry point for execution and therefore need to support more than one survey that requires validation.

**Comment by bcoca** [comment]
@privateip I had been thinking of that by adding 'per file' sections to the spec file (or a per file name spec file) . Either could specify the same structure as above (for main).

so either in spec.yml:
```
from:
 - <otherfile>:
     vars:
       ...
     config:
       ....
```

with the 'main' one being used as default if nothing else is set

to 'bypass' main you can just add empty entry:
```
vars:
  ...
config:
  ...
from:
  - <otherfile>: 
```

**Comment by bcoca** [comment]
another expansion on this i was thinking of is allow specifying defaults/vars files for the task, instead of vars/main.yml and default/main.yml ... but only if you don't specify from_vars/from_defaults in the caller.

**Comment by victorock** [comment]
@bcoca what about trying main.yaml then <filename> in meta and vars?
That would make the process transparent/simple for users, without breaking what we already have today.

Additionally, we could envision to have: 
1. meta action.
2. include_spec/import_spec.

Where:
1. Refresh default spec validator against defined variables (ansible_facts). Useful if for some reason variable values changed due to register/set_facts.
2. Explicitly define the spec to load/validate/revalidate against defined variables (ansible_facts).

**Comment by bcoca** [comment]
That is where it gets really tricky, for this stage, I think we should only deal with 'current role'.

Once you look at combinations of them you need a 'play spec'. Which is something i was also contemplating if this 'role spec' was successful.

**Comment by mwhahaha** [comment]
So if you do the validation at the start of each role, then you're only dealing with the 'current role' and would only be validating the current state at the start of execution so I think that's in line with what you were basing this on.  

For types I'm not saying support complex types, but support multiple simple types.  Being able to start locking down something to a string or an integer has to have a transition period where a user may have existing playbooks and are accidentally providing an integer when it should be a string:
```
---
my_var: 12
```
instead of 
```
---
my_var: '12'
```

**Comment by bcoca** [comment]
i was calling that 'complex', anything over the 'base type'

**Comment by mwhahaha** [comment]
Complex type for me would be additional schema validation of lists or json or something. What I wouldn't want is for everyone to just declare everything as strings or everything as optional.  I guess that is a work around for not supporting both but i'm not sure the benefit to truly being able to do any validation.  I would predict user frustration from breakages when it starts being implemented because folks aren't properly defining types.  My personal experience with this is that we saw this in the puppet3->puppet4 type conversions when folks started adding these types of constraints.  Being able to define multiple possible types prevents hard breakages during upgrades while allowing users a migration period.  There aren't really a bunch of types available via yaml so I'm not sure the added complexity of allowing for both an int/string or string/list combinations.

**Comment by chouseknecht** [comment]
Copying notes from #134...

After a bit deliberation we decided the following:

- Take Merged Approach -- survey and validation together, difference of two fields. Basically, bring this proposal together with #134
- Single file spec to put out for user input as POC
- Spec [will mirror this link](https://github.com/ansible/awx/blob/devel/awx/api/templates/api/job_template_survey_spec.md)
- Separate ansible-role (or like) command line tool
- Follow up meeting on implementation (@bcoca, @privateip, @chouseknecht, @alikins)
- Need to work on better naming of this stuff. (TBD)

**Comment by victorock** [comment]
What about also being able to use yang to create spec definitions?

**Comment by bcoca** [comment]
@victorock you can use anything you want to generate the spec, in the end its just a file. 

But I'm really unsure how yang works in this context though, since the spec represents 'role required vars' which normally feed into modules, that is what they need to conform to. The assumption would be that all actions also conform to the yang data.

**Comment by victorock** [comment]
@bcoca in yang you define the spec of supported/default values, including complex data like dict (leafs). As such any consumer willing to consume roles can rely on a standard spec to know what is the data structure.

**Comment by alikins** [comment]
Should ignore_errors on a include_role also cover the arg spec validation test?

ie, should a task with ignore_errors=true ignore validation errors

``` yaml

- include_role:
    name: some_role
    vars: 
       some_arg: some_invalid_value
  ignore_errors: true
```

or should it still cause the arg validation task to fail?

**Comment by bcoca** [comment]
its a task, but that would require the `ignore_errors` to be in the `apply:`  to continue with the role execution

**Comment by robertdebock** [comment]
It's exactly what [ansible-network has as an action_plugin](https://github.com/ansible-network/network-engine/blob/devel/action_plugins/validate_role_spec.py). Looks cool to me, but it's not a part of Ansible itself.

**Comment by tima** [comment]
Thanks @robertdebock. We have been engaged with the ansible-network team on this and have incorporated parts of their solution in this. The idea here is to make this part of Ansible itself once it's ready to ship.

**Comment by alikins** [comment]
https://github.com/ansible/ansible/pull/44983 has a implementation based on this proposal, ansible-engines validate_role_spec action plugin, and bits from a tower survey spec impl prototype.

**Comment by lwm** [comment]
Brilliant. This is very much needed.  Having a look at https://github.com/ansible/ansible/pull/44983/files now.

**Comment by lwm** [comment]
If anyone is interested in seeing how to move this forward, it will hopefully be discussed at the core meeting tomorrow at 7 UTC. See https://github.com/ansible/community/issues/418#issuecomment-456167944.

**Comment by ant31** [comment]
repost a comment from the PR  (https://github.com/ansible/ansible/pull/44983#issuecomment-438741499) here: 

 > I would suggest using OpenAPiv3 as the spec? there is already many tooling around this spec (to generate documentation, validation...). It's used by many other projects to define API and then already known by many users, contributed and tested and very well documented/specified.

IMO, it's an important design decision, I would really like to see arguments or discussion about following it or not.  

Also, what's the status of this feature ?

**Comment by decentral1se** [comment]
OK, I've been chasing this one a little in my spare time because it feels important. I first brought it to the core meeting which was a productive discussion: https://meetbot.fedoraproject.org/ansible-meeting/2019-01-29/ansible_core_public_irc_meeting.2019-01-29-19.01.log.html#l-16. TLDR; it was blocked based on some design discussion which was then unblocked in the meeting. The status is then that the ideal first step to make this arg spec possible is to remove the AnsibleModule argument specification validation from `basic.py` which is being done by @samdoran. They are submitting a number of PRs to make that happen which can be seen in the following card https://github.com/ansible/ansible/projects/30#card-13295996. All reviews are welcome! Once the argument validation logic is de-coupled from `AnsibleModule`, I will then take the minimum that we need from the nice implementation done in https://github.com/ansible/ansible/pull/44983 (as agreed in the meeting logs above) and we should have what we need! Once I get there, I'll start pinging for reviews and feedback. Hope that is clear!

**Comment by geerlingguy** [comment]
It's been a loooong time since any updates were posted on this issue, but I did want to mention today that @samdoran walked some people through an implementation using a yaml-based argspec inside `meta/` at the Ansible Contributor Summit today.

**Comment by bcoca** [comment]
This has been shipped in 2.11

**Comment by bendem** [comment]
Google search didn't bring anything, where is this documented, how can I use it today?

**Comment by bendem** [comment]
Answering myself, it's in the role doc : https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html#role-argument-validation
